src/com/MPISinglePortsCommunication.cpp

```diff
@@ -0,0 +1,207 @@
+#ifndef PRECICE_NO_MPI
+
+#include "MPISinglePortsCommunication.hpp"
+#include "utils/assertion.hpp"
+#include "utils/Parallel.hpp"
+#include "utils/Publisher.hpp"
+
+using precice::utils::Publisher;
+using precice::utils::ScopedPublisher;
+
+namespace precice
+{
+namespace com
+{
+MPISinglePortsCommunication::MPISinglePortsCommunication(std::string const &addressDirectory)
+    : _addressDirectory(addressDirectory)
+{
+  if (_addressDirectory.empty()) {
+    _addressDirectory = ".";
+  }
+}
+
+MPISinglePortsCommunication::~MPISinglePortsCommunication()
+{
+  TRACE(_isConnected);
+  closeConnection();
+}
+
+size_t MPISinglePortsCommunication::getRemoteCommunicatorSize()
+{
+  TRACE();
+  assertion(isConnected());
+  int size = -1;
+  MPI_Comm_remote_size(_communicators[0], &size);
+  return size;
+}
+
+void MPISinglePortsCommunication::acceptConnection(std::string const &acceptorName,
+                                                   std::string const &requesterName,
+                                                   int                acceptorRank)
+{
+  TRACE(acceptorName, requesterName);
+  assertion(not isConnected());
+
+  _isAcceptor = true;
+
+  MPI_Open_port(MPI_INFO_NULL, const_cast<char *>(_portName.data()));
+
+  const std::string addressFileName("." + requesterName + "-" + acceptorName + ".address");
+  Publisher::ScopedChangePrefixDirectory scpd(_addressDirectory);
+  ScopedPublisher                        p(addressFileName);
+  p.write(_portName);
+  DEBUG("Accept connection at " << _portName);
+
+  // Connect the first peer, s.t. we can exchange some information about the other side
+  MPI_Comm communicator;
```

Comment:
Please DRY your code a bit using anonymous namespaces.
This part of the code is essential a function from `(portName, acceptorRank) -> (requesterRank, requsterCommunicatorSize)`
```cpp
      MPI_Comm communicator;
      MPI_Comm_accept(const_cast<char *>(_portName.c_str()), MPI_INFO_NULL, 0, MPI_COMM_SELF, &communicator);
      DEBUG("Accepted connection at " << _portName);

      int    requesterRank      = -1;
      size_t requesterCommunicatorSize = 0;

      MPI_Recv(&requesterRank,             1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
      MPI_Recv(&requesterCommunicatorSize, 1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
      MPI_Send(&acceptorRank,              1, MPI_INT, 0, 42, communicator);
```

You could return a trivial intermediate such as:
```cpp
struct PeerReply {
    int rank;
    size_t communicatorSize;
};
```
This would make this code super readable and reduce the `MPI_Noise`

Reply:
The purpose of `setupCommunicatorSelf` is to setup the communicator. It is strictly dependant on the `_portName` and the result is used to register the communicator to `_communicators`. You also mentioned in your PM that there is no access to the member-logger, which is a problem.

So, lets just turn in into a private function: (Not tested)
```cpp
{
  TRACE(acceptorName, requesterName);
  assertion(not isConnected());

  _isAcceptor = true;

  MPI_Open_port(MPI_INFO_NULL, const_cast<char *>(_portName.data()));

  const std::string addressFileName("." + requesterName + "-" + acceptorName + ".address");
  Publisher::ScopedChangePrefixDirectory scpd(_addressDirectory);
  ScopedPublisher                        p(addressFileName);
  p.write(_portName);
  DEBUG("Accept connection at " << _portName);
  // Setup first peer
  const auto firstCommSize = setupCommunicator(acceptorRank);
  
  // Connect all other peers
  for (size_t i = 1; i < firstCommSize; ++i) {
      const auto commSize = setupCommunicator(acceptorRank);
      assertion(firstCommSize == commSize, "Inconsistent communicator sizes!")
  }
  
  _isConnected = true;
}
```
With:
```cpp
size_t MPISinglePortsCommunication::setupCommunicator(int acceptorRank) {
  MPI_Comm communicator;
  MPI_Comm_accept(const_cast<char *>(_portName.c_str()), MPI_INFO_NULL, 0, MPI_COMM_SELF, &communicator);
  DEBUG("Accepted connection at " << _portName);

  int    requesterRank      = -1;
  size_t requesterCommunicatorSize = 0;

  // Exchange information to which rank I am connected and which communicator size on the other side
  MPI_Recv(&requesterRank,             1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
  MPI_Recv(&requesterCommunicatorSize, 1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
  MPI_Send(&acceptorRank,              1, MPI_INT, 0, 42, communicator);
  
  CHECK(requesterCommunicatorSize > 0, "Requester communicator size has to be > 0!");
  _communicators[requesterRank] = communicator;
  return requesterCommunicatorSize;
}
``` 
This would also allow logging using the class member.
I am not sure about the name of the function though.

Reply:
The point I am trying to make is that this code handles MPI orchestration, which is hard to debug as problems don't result in compile time errors and lead to runtime deadlocks.
Code duplication in the core of such a function is in my opinion absolutely unacceptable.
That is the core message of my review.
The code examples I provided are only suggestions of how one might DRY the code.

Regarding your feedback:

> * Now that it is a member and modifies the class' state, it has none of the elegance and clearly defined scope of a function (in mathematical sense)

Logging and MPI will always imply a state on this code. I agree that a clean separation would be favourable, but I don't see a possibility here.

> * It deludes the reader into believing that this way of communicator setup is the one default/standard way done in that class, while in fact it is only used in one of four functions that create communicator.

This is an implementation detail, which is what documentation and private functions are for.
That said, the difficult part would be to pick an appropriate name for said function.

> * The net win in terms of lines of code is minimal and the it increases complexity.

Firstly, I don't think that LOC is the appropriate metric here. Secondly, I think that this significantly decreases complexity.
From my perspective it is easier to reason about a small member function than having to track how variables change throughout a long member function that even involves a loop.
Isolating the comm-setup leads to smaller functions with fewer variables and state changes reduced to member variables only.

Reply:
As an alternative to extracting the section, a `do while` loop could be used to combine the first iteration with the rest:
```cpp
{
  TRACE(acceptorName, requesterName);
  assertion(not isConnected());

  _isAcceptor = true;

  MPI_Open_port(MPI_INFO_NULL, const_cast<char *>(_portName.data()));

  const std::string                      addressFileName("." + requesterName + "-" + acceptorName + ".address");
  Publisher::ScopedChangePrefixDirectory scpd(_addressDirectory);
  ScopedPublisher                        p(addressFileName);
  p.write(_portName);

  size_t peerCurrent = 0; // The current Peer to connect to
  size_t peerCount   = 0; // The total count of peers (initialized in the first iteration)
  do {
    MPI_Comm communicator;
    MPI_Comm_accept(const_cast<char *>(_portName.c_str()), MPI_INFO_NULL, 0, MPI_COMM_SELF, &communicator);
    DEBUG("Accepted connection at " << _portName << " for Peer " << peerCurrent);

    // Handshake with Peer
    int    requesterRank             = -1;
    size_t requesterCommunicatorSize = 0;

    MPI_Recv(&requesterRank, 1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
    MPI_Recv(&requesterCommunicatorSize, 1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
    MPI_Send(&acceptorRank, 1, MPI_INT, 0, 42, communicator);

    // Check for consistency
    CHECK(requesterCommunicatorSize > 0, "Requester communicator size has to be > 0!");
    CHECK(requesterCommunicatorSize == _communicators.size(),
          "Requester communicator sizes are inconsistent!");
    CHECK(_communicators.count(requesterRank) == 0,
          "Duplicate request to connect by same rank (" << requesterRank << ")!");

    // Register the communicator for this peer connection
    _communicators[requesterRank] = communicator;

    // Initialize the count of peers to connect to
    if (peerCurrent == 0) {
        peerCount = requesterCommunicatorSize;
        DEBUG("Connections per Peer: " << peerCount);
    }
  } while (++peerCurrent < peerCount);

  _isConnected = true;
}
```

Reply:
I like that way much more! Adapted it a bit
```cpp
void MPISinglePortsCommunication::acceptConnection(std::string const &acceptorName,
                                                   std::string const &requesterName,
                                                   int                acceptorRank)
{
  TRACE(acceptorName, requesterName);
  assertion(not isConnected());

  _isAcceptor = true;

  MPI_Open_port(MPI_INFO_NULL, const_cast<char *>(_portName.data()));

  const std::string addressFileName("." + requesterName + "-" + acceptorName + ".address");
  Publisher::ScopedChangePrefixDirectory scpd(_addressDirectory);
  ScopedPublisher                        p(addressFileName);
  p.write(_portName);

  size_t peerCurrent = 0; // current peer to connect to
  size_t requesterCommunicatorSize = 0;

  do {
    // Connection
    MPI_Comm communicator;
    MPI_Comm_accept(const_cast<char *>(_portName.c_str()), MPI_INFO_NULL, 0, MPI_COMM_SELF, &communicator);
    DEBUG("Accepted connection at " << _portName << " for peer " << peerCurrent);

    int requesterRank = -1;
    // Exchange information to which rank I am connected and which communicator size on the other side
    MPI_Recv(&requesterRank,             1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
    MPI_Recv(&requesterCommunicatorSize, 1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
    MPI_Send(&acceptorRank,              1, MPI_INT, 0, 42, communicator);

    CHECK(requesterCommunicatorSize > 0,
          "Requester communicator size has to be > 0!");
    CHECK(_communicators.count(requesterRank) == 0,
          "Duplicate request to connect by same rank (" << requesterRank << ")!");
    
    _communicators[requesterRank] = communicator;

  } while (++peerCurrent < requesterCommunicatorSize);
  
  _isConnected = true;
}
```
I took the same scheme for MPIPortsCommunication.